### DISCUSSION TITLE
SinglePhaseFluidProperties error with NaKFluidProperties

### DISCUSSION THREAD

**Discussion by garcs2** [idaholab/moose#26791] [Q&A Modules: Thermal Hydraulics] [answered]
Hello,

I am trying to use THM to predict the outlet conditions of a sodium-potassium cooled thermal microreactor. I am following directly from the single phase flow THM tutorial, and have used the conjugate heat transfer example as a template to begin with. I've effectively adjusted TH properties for what I'm modeling, including the fluid. I use NaKFluidProperties to define the fluid and get the following error when running on lemhi:
```
The following error occurred in the object "nak", of type "NaKFluidProperties".

nak: virtual void SinglePhaseFluidProperties::p_from_v_e(libMesh::Real, libMesh::Real, libMesh::Real&, libMesh::Real&, libMesh::Real&) const derivatives not implemented.
```
I'm confused on whether this error is telling me that the variable p_from_v_e isn't implemented in my own input file or if that has not been implemented in general? In going through NaKFluidProperties documentation I didn't find that variable to input so I'm confused what the solution moving forward is. Thanks in advance!

**Comment by GiudGiud** [comment, marked as answer]
Hello

This is not an input file problem, it s a missing implementation in the source code for the fluid property.
If you look at the NaKFluidProperties.h header file, you ll see that this routine does not exist. 

To move forward you will need to implement it. This can be challenging for these weakyl/incompressible fluids as (v, e) variables are both only correlated to temperature and not pressure. 

Guillaume

**Reply by garcs2** [reply]
Okay that makes sense, thanks so much for the response!

**Reply by GiudGiud** [reply]
Btw if you get them implemented and they work fine we would be glad to see these routines contributed to the module.

**Comment by BrankoK89** [comment]
Hi, I wanted to ask my question here as the issues appear to be related. I am trying to model 1D TH for a fuel salt channel (by modifying the available tutorial https://cardinal.cels.anl.gov/tutorials/openmc_fluid.html#n1) and I am using     type = FlibeFluidProperties, but I am getting the error: 
The fluid properties class 'FlibeFluidProperties' has not implemented the method below, which computes derivatives of fluid properties with regards to the flow variables. If your application requires this method, you must either implement it or use a different fluid properties  class.                                                                                                                                                                                                                                                                              
virtual void SinglePhaseFluidProperties::c_from_v_e(libMesh::Real, libMesh::Real, libMesh::Real&, libMesh::Real&, libMesh::Real&) const                                                                                                                                                                                                                                      
You can avoid this error by neglecting the unimplemented derivatives of fluid properties by setting the 'allow_imperfect_jacobians' parameter 

Setting the allow_imperfect_jacobians=True did not alleviate the issue: The fluid properties class 'FlibeFluidProperties' has not implemented the method below. If your application requires this method, you must either implement it or use a different fluid properties class. 

So it probably means the same thing, the method does not exist. I would really appreciate it if you would advise on a proper way to then model flibe properties as I am very new to MOOSE and thm.

**Reply by GiudGiud** [reply]
> So it probably means the same thing, the method does not exist. I would really appreciate it if you would advise on a proper way to then model flibe properties as I am very new to MOOSE and thm

You will need to figure out how to implement c_from_v_e in FLiBeFluidProperties
The speed of sound might be constant for FLiBe actually, so it might be easy to implement. 
I would recommend searching the references mentioned in the object to see if they contain this addition c fluid property

**Reply by BrankoK89** [reply]
Thank you for the reply. I have never done this so it took some time to digest. So I could go to my Cardinal folder and find MOOSE folder in there and find FlibeFluidProperties.C (C++ source code)  and FlibeFluidProperties.h and adjust them by adding this property? 

Would something like this work:
in .h:
```
/**
 * Compute speed of sound [m/s] from specific volume and internal energy
 */
virtual Real c_from_v_e(Real v, Real e) const override;
```
Then in .C:
```
Real
FlibeFluidProperties::c_from_v_e(Real v, Real e) const
{
  Real p, dp_dv, dp_de;

  p_from_v_e(v, e, p, dp_dv, dp_de);

  Real c2 = -v * v * dp_dv;

  if (c2 < 0.0)
    mooseError("Computed c^2 < 0, check EOS consistency!");

  return std::sqrt(c2);
}
```

**Reply by GiudGiud** [reply]
> Would something like this work:

yes though isnt it supposed to be dp_dv at constant entropy instead of at constant specific energy?

**Reply by BrankoK89** [reply]
Yes, it is, but there seems to be no formulation for constant entropy in that object. Joshua Hansel told me this: currently there's no formulation for the entropy, so there's no way to derive the isentropic derivative there. However, it seems there is at least (dp/drho)_T, which might be a decent approximation. However, I wasn't able to find (dp/drho)_T in that object neither, but I did find stuff at constant specific energy. Is that an OK approximation?

**Reply by GiudGiud** [reply]
probably not, v and e are tied to temperature. (dp_drho)_T should be computable (from drho_dp and drho_dT), you just need to convert v,e to p,T and keep track of the derivatives

**Reply by BrankoK89** [reply]
OK, thank you for the explanation. This was rather tough for me as I am new to it, but would you tell me if it is good:
```
Real 
FlibeFluidProperties::c_from_v_e(Real v, Real e) const
{
  Real p, dp_dv, dp_de;
  p_from_v_e(v, e, p, dp_dv, dp_de); // get pressure and derivatives at constant e

  Real e_val, de_dp, de_dT;
  e_from_p_T(p, T_from_v_e(v,e), e_val, de_dp, de_dT); // e and derivatives at constant T

  Real v_val, dv_dp, dv_dT;
  v_from_p_T(p, T_from_v_e(v,e), v_val, dv_dp, dv_dT);

  // Compute (de/dv)_T = (de/dp)_T / (dv/dp)_T
  Real de_dv_T = de_dp / dv_dp;

  // Compute (dp/dv)_T using chain rule
  Real dp_dv_T = dp_dv + dp_de * de_dv_T;

  Real c2 = -v * v * dp_dv_T;

  if (c2 < 0.0)
    mooseError("Computed c^2 < 0, check EOS consistency!");

  return std::sqrt(c2);
}
```

**Reply by BrankoK89** [reply]
What do you think about this code for implementing e_from_v_h (another error): 
```
Real
FlibeFluidProperties::e_from_v_h(Real v, Real h) const
{
  Real e = h; // good initial guess
  for (unsigned int i = 0; i < 10; ++i)
  {
    Real p = p_from_v_e(v, e);
    Real R = e + p * v - h;
    Real dp_dv, dp_de, dummy;
    p_from_v_e(v, e, dummy, dp_dv, dp_de);
    Real dR_de = 1.0 + dp_de * v;
    Real delta_e = -R / dR_de;
    e += delta_e;
    if (std::abs(delta_e) < 1e-12)
      break;
  }
  return e;
}
```

**Reply by GiudGiud** [reply]
there is a utility in the FP module for doing Newton's method. You should not need to re-code it. 
but yes it's a good idea

**Reply by BrankoK89** [reply]
OK, that helps! How do I invoke the Newton's method utility? I tried doing:
Real
FlibeFluidProperties::e_from_v_h(Real v, Real h) const
{
  FluidPropertiesUtils::NewtonRaphson<Real> solver;

  auto residual = [&](Real e_trial)
  {
    return h - (e_trial + p_from_v_e(v, e_trial) * v);
  };

  auto derivative = [&](Real e_trial)
  {
    Real dummy, dp_de, dp_dv;
    p_from_v_e(v, e_trial, dummy, dp_dv, dp_de);
    return -(1.0 + v * dp_de);
  };

  return solver.solve(residual, derivative, h, _tolerance, _max_newton_its);
}
But I keep getting the error when I try to rebuild: 
/home/branko89/cardinal/contrib/moose/modules/fluid_properties/src/fluidproperties/FlibeFluidProperties.C:387:25: error: 'NewtonRaphson' is not a member of 'FluidPropertiesUtils'
  387 |   FluidPropertiesUtils::NewtonRaphson<Real> solver;

**Reply by GiudGiud** [reply]
see fluid_properties/include/utils/NewtonInversion.h for the function definitions
